File: LeetCode/Medium/1379_Find_a_Corresponding_Node_of_a_Binary_Tree_in_a_Clone_of_That_Tree/Solution.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

## 2022-05-16
class Solution:
    def getTargetCopy(self, original: TreeNode, cloned: TreeNode, target: TreeNode) -> TreeNode:
        if not original:
            return None
        
        if original == target:
            return cloned
        
        left = self.getTargetCopy(original.left, cloned.left, target)
        right = self.getTargetCopy(original.right, cloned.right, target)
        
        return left if left else right
        

File: LeetCode/Medium/1379_Find_a_Corresponding_Node_of_a_Binary_Tree_in_a_Clone_of_That_Tree/Solution.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

// 2022-05-13
class Solution {
    public final TreeNode getTargetCopy(final TreeNode original, final TreeNode cloned, final TreeNode target) {
        if (original == target) return cloned;
        
        TreeNode answer = null;
        if (original.left != null)
            answer = getTargetCopy(original.left, cloned.left, target);
        if (answer == null && original.right != null)
            answer = getTargetCopy(original.right, cloned.right, target);
        return answer;
    }
}